service/raft/raft_group0_client.hh

```diff
@@ -53,6 +53,8 @@ public:
     // Until the upgrade procedure finishes, we will perform operations such as schema changes using the old way,
     // but still pass the guard around to synchronize operations with the upgrade procedure.
     bool with_raft() const;
+
+    explicit operator bool() const { return bool(_impl); }
```

Comment:
Ugh. Making guard an optional object makes reasoning about it much harder.

Now whenever reviewing code that takes a guard I need to think --- but what if it's disengaged? Can it be disengaged?

Why not use `std::optional<group0_guard>` instead? This basically imposes the complexity of reasoning with `optional`s on every code path that uses `group0_guard`, even when it's non-optional!

Reply:
I think this is a big mistake.

Reply:
The `_impl` was done for pimplification and only  for that

Reply:
Ideally, having `group0_guard` object in hand should give a type-level guarantee that it's safe to do linearizable group 0 operation and concurrent modifications will be detected.

Using a guard after a move... is a use-after-move, it's a bug.

Reply:
Well, the guard is still non-optional, the operator bool() checks if it was moved-from. Using a moved-from guard is still incorrect.

Reply:
But calling public interface on a moved-from object is basically using that moved-from object

Maybe we could find better tools to deal with such cases, something that would protect us on type level...